Improve Android build

RandomShaper opened the pull request:
**Edit:** Keeping just the updated, consolidated text. If this PR gets merged, I can update the docs.

**Overall note: All this proposal is based on the assumption that the Android build is only used for generating the export templates as reliably as possible. Moreover, I believe that makes the Android build handier. If in some future tools-enabled Android builds were needed because of a widespread desire of using Godot under Android netbooks or so, some things would need to be changed.**
## Android/build component versions
- Upgrade Android plugin for Gradle to 2.1.0
- Downgrade NDK platform to 14 to match minSdkVersion.
## Smarter build and packaging of the Godot export template
### SCons phase

When the SCons phase of the Android build is done the generated .so file is moved to the directory relevant for the Gradle phase to pick it for the combination of arch and build type (release or debug). For instance, `platform/android/java/libs/debug/armeabi-v7a`.

Moving is performed instead of copying or symlinking to avoid polluting the `bin` directory with intermediate files, because the final binaries for Android are the APKs. The only scenario (and is a theoretical one) in which the .`so` would be kept at `bin` is that of building for an arch unknown to Android. And in that case, the SCons script would print a warning message.

Notes:
- ARMv7 both with or without NEON will be copied to the same `armeabi-v7a` dir (the newest wins).
- Both _debug_ and _debug_release_ targets will get copied to the same `debug` dir (the newest wins). Release is unambiguous.
### Gradle phase

Because of all that is done at the SCons level, Gradle will be able to generate both debug and release APKs containing libraries for x86, ARMv6 and/or ARMv7. In the case of the debug APK it can contain the handy release_debug or a full debug build if needed.

Furthermore, Gradle will generate the export template APKs with their correct naming (`android_debug.apk` and `android_release.apk`) and at the correct location at the top-level `bin` directory.

Also the Gradle script disables explicitly singing and zip-aligning of the APKs to save time since those operations must be done at export time then the final game APK is generated.

RandomShaper commented:
If people find this work useful I'll start working on the TO-DOs (especially docs) ASAP.

vnen commented:
Does this supersedes (or is related) to #4729? @volzhs might want to have a look.

volzhs commented:
What I did is supporting target sdk 23.
Some of @RandomShaper's changes was done during I fixing the issue.

RandomShaper commented:
Yes, I think our PRs overlap in tweaking certain version numbers and API levels. Nonetheless, I think they could be nicely merged

@volzhs please tell me if you could take advantage of the last Android plugin for Gradle (2.1) so we both use the same.

On merging I think my 'checkReleaseBuilds false' should disappear in favor of your better handling of lints.

akien-mga commented:
I just merged @volzhs's PR, so as expected there are now merge conflicts. Could you rebase yours on top of the current master branch?

volzhs commented:
@RandomShaper I'm not sure what kind of advantage I can take with gradle 2.1.
But I compiled with supporting android api 23 and gradle 2.1 and tested it api 19 and 23 phone.
It works fine. :)

slapin commented:
Please copy ARMv6 one too, then I will be able to test GLES on some of 3 year old devices :)
(p.s. Android 4.0)

RandomShaper commented:
3 years doesn't seem too old to exclude them. Could you name one device?

I'm curious about OpenGL ES 2.0 powered ARMv6 devices (thus with VFP),

slapin commented:
This is specialized device, I currently did not integrate hardware GL
libraries there yet,
but it does have SGX5 hardware in SoC.

As for example device I think the most common is RaspberryPi,
as I see, but I think there are others still used.

On Thu, Jun 9, 2016 at 2:34 AM, Pedro J. Estébanez <notifications@github.com
> wrote:

> 3 years doesn't seem too old to exclude them. Could you name one device?
>
> I'm curious about OpenGL ES 2.0 powered ARMv6 devices (thus with VFP),
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/5077#issuecomment-224762663>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAX0_wccLrgpF7FrgpN130I_5iOL5zIks5qJ1GSgaJpZM4IvbuA>
> .
>

slapin commented:
Actually there was a lot of devices armv6 + gles2 made in 2011, with some
made even in 2013,
but I have what I have, it is 1.2GHz single-core armv6 SoC, 512M RAM, used
as kind of street set top box
now retargeted for some other usage.
Not that it is very important, it just have ported Android 4.0 there (was
2.3 originally),
so it would be fun to test if Godot will run there.


On Thu, Jun 9, 2016 at 3:27 AM, Sergey Lapin <slapinid@gmail.com> wrote:

> This is specialized device, I currently did not integrate hardware GL
> libraries there yet,
> but it does have SGX5 hardware in SoC.
>
> As for example device I think the most common is RaspberryPi,
> as I see, but I think there are others still used.
>
> On Thu, Jun 9, 2016 at 2:34 AM, Pedro J. Estébanez <
> notifications@github.com> wrote:
>
>> 3 years doesn't seem too old to exclude them. Could you name one device?
>>
>> I'm curious about OpenGL ES 2.0 powered ARMv6 devices (thus with VFP),
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/5077#issuecomment-224762663>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAAX0_wccLrgpF7FrgpN130I_5iOL5zIks5qJ1GSgaJpZM4IvbuA>
>> .
>>
>
>

RandomShaper commented:
Well, unless somebody founds errors (technical or logical) about all these, I think I've completed this task.

volzhs commented on platform/android/build.gradle.template:
```diff
@@ -26,11 +26,11 @@ android {
-	compileSdkVersion 23
+	compileSdkVersion 19
 	buildToolsVersion "23.0.3"
-	useLibrary 'org.apache.http.legacy'
```
```
compileSdkVersion 23
useLibrary 'org.apache.http.legacy'
```
why are these lines changed back?
it needs for resolving issue #4711

RandomShaper replied:
Lost in merging. My fault. Line pushed back.

volzhs commented on platform/android/build.gradle.template:
```diff
@@ -26,9 +26,10 @@ android {
 	}
 
-	compileSdkVersion 23
+	compileSdkVersion 19
```
I don't know how to refer multiple lines with github line note.
would you revert it to 23 too?

RandomShaper replied:
Reverted.

akien-mga commented:
@RandomShaper Would you be able to squash together all commits, or at least group them in related chunks to keep only a few of them? You have a pretty good practice regarding making small and well described commits, but 18 commits in the end for a relatively small overall patch is a bit too much compared to other commits in this repo: https://patch-diff.githubusercontent.com/raw/godotengine/godot/pull/5077.patch :)

RandomShaper commented:
I've added a couple of ignores corresponding to directories the old Android Eclipse-based dev tools generated.

And I've put everything together in a single commit.

RandomShaper committed:
Make Android build smarter (SCons + Gradle)
Upgrade Gradle and Android plugin for Gradle
Disable all signing and zip-aligning for the export templates
Give correct names to generated APKs
Put .so files built by SCons right where Gradle has to pick them according to arch & build type
Downgrade NDK platform to 14 to match minSdkVersion

akien-mga commented:
Thanks! If I can be annoying a bit more :-P Would you mind amending the commit log (`git commit --amend`) to add some more information about what it does?

Typically a git log contains a title (first line), and a description (all the contents that comes after a newline), so it could be:

```
Make Android build smarter (SCons + Gradle)

With an intricate system of tubes and heat exchangers, this commit achieves a higher
efficiency in the frying of French (or freedom) fries, using the waste heat as an external
power source to trigger nightly builds of Godot binaries.
```

RandomShaper commented:
Have a look now. I think I've almost got the workflow. :)

akien-mga commented:
That's perfect, thanks! :)

Hinsbart commented:
@RandomShaper I have a problem building templates for android. When I do changes to the c++ side, they're not reflected when I run the resulting app.
Even though scons compiles the libgodot_android.so and it seems allright from what I can tell... :/
Could that be related to this PR?

(Changes to the Java code work as expected btw)

RandomShaper commented:
Yeah. Some people is having issues.

Please tell me the complete command you used for building so I can check
whether I missed some case.

Anyway, I'm on my honeymoon so I won't be able to do it before a few days.
El 12/7/2016 9:22 a. m., "Andreas Haas" <notifications@github.com> escribió:

> @RandomShaper <https://github.com/RandomShaper> I have a problem building
> templates for android. When I do changes to the c++ side, they're not
> reflected when I run the resulting app.
> Even though scons compiles the libgodot_android.so and it seems allright
> from what I can tell... :/
> Could that be related to this PR?
>
> (Changes to the Java code work as expected btw)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/5077#issuecomment-232062888>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALQCtuQjPp2HwG0De41iQpiToyaludtzks5qU6M8gaJpZM4IvbuA>
> .
>
